lib/prometheus-parser/src/lib.rs

```diff
@@ -233,12 +306,157 @@ pub fn group_metrics(input: &str) -> Result<Vec<MetricGroup>, ParserError> {
     Ok(groups)
 }
 
+#[derive(Default)]
+struct MetricGroupSet(IndexMap<String, GroupKind>);
+
+impl MetricGroupSet {
+    fn get_group<'a>(&'a mut self, name: &str) -> (usize, &'a String, &'a mut GroupKind) {
+        let len = name.len();
+        let name = if self.0.contains_key(name) {
+            name
+        } else if name.ends_with("_bucket") && self.0.contains_key(&name[..len - 7]) {
+            &name[..len - 7]
+        } else if name.ends_with("_sum") && self.0.contains_key(&name[..len - 4]) {
+            &name[..len - 4]
+        } else if name.ends_with("_count") && self.0.contains_key(&name[..len - 6]) {
+            &name[..len - 6]
+        } else {
+            self.0
+                .insert(name.into(), GroupKind::new(MetricKind::Untyped));
+            name
+        };
+        self.0.get_full_mut(name).unwrap()
+    }
+
+    fn insert_metadata(&mut self, name: String, kind: MetricKind) -> Result<(), ParserError> {
+        if let Some(group) = self.0.get(&name) {
+            if !group.matches_kind(kind) {
+                return Err(ParserError::MultipleMetricKinds { name });
+            }
```

Comment:
So in the `else` case we don't insert anything but return `Ok`. Is this correct?

Reply:
Right, because the metadata already exists and is the right kind. I'll add a comment to that effect.